### PR TITLE
Consensus Round as Key in YacStorages

### PR DESCRIPTION
**Pull request by Akvinikym**
### Description of the Change

This PR introduces consensus Round, which is used as a key in Yac storages, allowing new BFT algorithms to be used.

### Benefits

One more step to new ordering service.

### Possible Drawbacks 

None.

### CODE REVIEW

File: irohad/ordering/on_demand_os_transport.hpp [hunk 1]
```diff
@@ -6,15 +6,14 @@
 #ifndef IROHA_ON_DEMAND_OS_TRANSPORT_HPP
 #define IROHA_ON_DEMAND_OS_TRANSPORT_HPP
 
-#include <cstdint>
 #include <memory>
 #include <tuple>
```
Comment (lebdron) [inline irohad/ordering/on_demand_os_transport.hpp]: `<tuple>` should be moved to `round.hpp`, see `std::tie`

File: irohad/consensus/round.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        boost::hash_combine(seed, val.block_round);
+        boost::hash_combine(seed, val.reject_round);
+        return seed;
+      }
+    };
+
+    /**
+     * Type of proposal and round this proposal was voted for in
+     */
+    using ProposalWithRound =
```
Comment (lebdron) [inline irohad/consensus/round.hpp]: These are unrelated to round, please move ProposalWithRound to ordering gate interface, and replace `pair` with struct. No need to use shared pointer with Round.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 3]
```diff
@@ -49,13 +49,17 @@ namespace iroha {
...
-            [this](auto block) { this->vote(block); });
+            [this](const auto &block_with_round) {
+              this->vote(block_with_round);
+            });
       }
 
-      void YacGateImpl::vote(std::shared_ptr<shared_model::interface::Block> block) {
-        auto hash = hash_provider_->makeHash(*block);
+      void YacGateImpl::vote(
+          const consensus::BlockWithRound &block_with_round) {
```
Comment (lebdron) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: Please check https://github.com/hyperledger/iroha/pull/1731 . Voting will require <Round, Proposal, Block>

File: irohad/consensus/round.hpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        std::pair<std::shared_ptr<shared_model::interface::Proposal>,
+                  std::shared_ptr<Round>>;
+
+    /**
+     * Type of block and round this block was voted for in
+     */
+    using BlockWithRound =
```
Comment (lebdron) [inline irohad/consensus/round.hpp]: This type should not be required at all.

File: irohad/network/peer_communication_service.hpp [hunk 5]
```diff
@@ -49,12 +50,11 @@ namespace iroha {
...
        * Event is triggered when proposal arrives from network.
-       * @return observable with Proposals.
-       * (List of Proposals)
+       * @return observable with Proposals and rounds, in which proposals
+       * appeared (List of Proposals)
        */
-      virtual rxcpp::observable<
-          std::shared_ptr<shared_model::interface::Proposal>>
-      on_proposal() const = 0;
+      virtual rxcpp::observable<consensus::ProposalWithRound> on_proposal()
```
Comment (lebdron) [inline irohad/network/peer_communication_service.hpp]: Is it required to change PCS in this pull request?

File: irohad/consensus/yac/yac_hash_provider.hpp [hunk 6]
```diff
@@ -36,30 +37,41 @@ namespace iroha {
...
          */
-        std::string proposal_hash;
+        Round vote_round_;
 
         /**
-         * Hash computed from block;
+         * Contains hashes of proposal and block, for which peer voted
          */
-        std::string block_hash;
+        struct VoteHashes {
```
[the remark this answers is not in the document]
Comment (Akvinikym) [reply irohad/consensus/yac/yac_hash_provider.hpp]: IMHO it makes the structure of YacHash more clear and structured, sorry for tautology

File: irohad/consensus/yac/transport/yac_pb_converters.hpp [hunk 7]
```diff
@@ -24,8 +24,12 @@ namespace iroha {
           proto::Vote pb_vote;
 
           auto hash = pb_vote.mutable_hash();
-          hash->set_block(vote.hash.block_hash);
-          hash->set_proposal(vote.hash.proposal_hash);
+          auto hash_round = hash->mutable_vote_round();
```
Comment (l4l) [inline irohad/consensus/yac/transport/yac_pb_converters.hpp]: Maybe extract function for round (de-)serialization? So these functions will correspond to related proto types

File: irohad/consensus/yac/storage/yac_common.hpp [hunk 8]
```diff
@@ -29,9 +29,9 @@ namespace iroha {
 
       struct VoteMessage;
 
-      using ProposalHash = decltype(YacHash::proposal_hash);
+      using ProposalHash = decltype(YacHash::vote_hashes_.proposal_hash);
```
Comment (l4l) [inline irohad/consensus/yac/storage/yac_common.hpp]: Imo it's better define types here and then reuse them in YacHash definition. With that approach ctor types can be also be "generic"

File: irohad/consensus/yac/yac_hash_provider.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/consensus/yac/yac_hash_provider.hpp]: Quite arguable, but ok if you'd like it more

File: irohad/consensus/round.hpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+      BlockRoundType block_round;
+      RejectRoundType reject_round;
+
+      Round() = default;
+
+      Round(BlockRoundType block_r, RejectRoundType reject_r)
+          : block_round{block_r}, reject_round{reject_r} {}
+
+      bool operator<(const Round &rhs) const {
+        return std::tie(block_round, reject_round)
```
Comment (lebdron) [inline irohad/consensus/round.hpp]: This should be in a cpp file due to tuple instantiation.

File: irohad/consensus/yac/storage/impl/yac_proposal_storage.cpp [hunk 10]
```diff
@@ -25,35 +25,33 @@ namespace iroha {
...
+                                       block_storage.getStorageKey();
+                                   return storage_key == store_hash;
+                                 });
         if (iter != block_storages_.end()) {
           return iter;
         }
         // insert and return new
         return block_storages_.emplace(block_storages_.end(),
-                                       YacHash(proposal_hash, block_hash),
+                                       std::move(store_hash),
```
Comment (lebdron) [inline irohad/consensus/yac/storage/impl/yac_proposal_storage.cpp]: Moving a const ref is not useful, move call can be removed.

File: irohad/consensus/yac/storage/impl/yac_vote_storage.cpp [hunk 11]
```diff
@@ -57,21 +57,20 @@ namespace iroha {
...
       }
 
-      bool YacVoteStorage::isHashCommitted(ProposalHash hash) {
-        auto iter = getProposalStorage(std::move(hash));
+      bool YacVoteStorage::isCommitted(const Round &round) {
+        auto iter = getProposalStorage(round);
         if (iter == proposal_storages_.end()) {
           return false;
         }
         return bool(iter->getState());
```
Comment (lebdron) [inline irohad/consensus/yac/storage/impl/yac_vote_storage.cpp]: Call to bool looks redundant since bool is specified as the return type.

File: irohad/consensus/yac/storage/yac_block_storage.hpp [hunk 12]
```diff
@@ -105,18 +104,18 @@ namespace iroha {
         bool uniqueVote(VoteMessage &vote);
 
         /**
-         * Verify that vote has same proposal and
-         * blocks hashes with storage
+         * Verify that vote has the same round attached as the storage
```
Comment (lebdron) [inline irohad/consensus/yac/storage/yac_block_storage.hpp]: It compares the whole hash, not only round. Please reflect this in the docstring.

File: irohad/consensus/yac/storage/yac_vote_storage.hpp [hunk 13]
```diff
@@ -104,28 +104,28 @@ namespace iroha {
...
-        bool isHashCommitted(ProposalHash hash);
+        bool isCommitted(const Round &round);
 
         /**
-         * Method provide state of processing for concrete hash
-         * @param hash - target tag
-         * @return value attached to parameter's hash. Default is false.
+         * Method provide state of processing for concrete proposal/block
+         * @param round, in which that proposal/block is being voted
+         * @return value attached to parameter's round. Default is false.
```
Comment (lebdron) [inline irohad/consensus/yac/storage/yac_vote_storage.hpp]: false -> kNotSentNotProcessed

File: irohad/consensus/yac/storage/impl/yac_common.cpp [hunk 14]
```diff
@@ -33,15 +33,15 @@ namespace iroha {
         auto first = votes.at(0);
         return std::all_of(
             votes.begin(), votes.end(), [&first](const auto &current) {
-              return first.hash.proposal_hash == current.hash.proposal_hash;
+              return first.hash.vote_round == current.hash.vote_round;
             });
       }
 
       boost::optional<ProposalHash> getProposalHash(
```
Comment (lebdron) [inline irohad/consensus/yac/storage/impl/yac_common.cpp]: This method is used only in tests now, and previously it was testing that round is the same. Please remove the method, or change its behavior to return the round.

File: irohad/consensus/yac/storage/impl/yac_common.cpp [hunk 15]
```diff
@@ -25,29 +25,27 @@ namespace iroha {
...
             });
       }
 
-      boost::optional<ProposalHash> getProposalHash(
-          const std::vector<VoteMessage> &votes) {
-        auto &&hash = getHash(votes);
-        if (hash) {
-          return (*hash).proposal_hash;
+      boost::optional<Round> getKey(const std::vector<VoteMessage> &votes) {
+        if (not sameKeys(votes)) {
```
Comment (l4l) [inline irohad/consensus/yac/storage/impl/yac_common.cpp]: `.. or votes.size() == 0`

Comment (lebdron) [reply irohad/consensus/yac/storage/impl/yac_common.cpp]: No, because vote messages can contain different keys.

Comment (Akvinikym) [reply irohad/consensus/yac/storage/impl/yac_common.cpp]: This is checked inside `sameKeys()`
